lib/Transforms/IPO/StructFieldCacheAnalysis.cpp

```diff
@@ -36,30 +46,672 @@ class StructFieldCacheAnalysisPass : public ModulePass {
+  Op = U->getOperand(2);
+  assert(isa<Constant>(Op));
+  auto* Index = dyn_cast<Constant>(Op);
+  auto Offset = (FieldNumType)Index->getUniqueInteger().getZExtValue();
+  assert(Offset < NumElements);
+  //TODO: ignore indices after this one. If there's indices, the field has to be an array or struct
+  return Offset+1; // return field number starting from 1
+}
+
+void StructFieldAccessInfo::addFieldAccessFromGEP(const User* U)
+{
+  DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Analyze all users of GEP: " << *U << "\n");
+  assert(isa<GetElementPtrInst>(U) || isa<GEPOperator>(U));
+  auto FieldLoc = calculateFieldNumFromGEP(U);
+  if (FieldLoc == 0)
+    return;
+  for (auto *User : U->users()){
+    DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Check user of " << *U << ": " << *User << "\n");
+    if (isa<Instruction>(User)){
+      auto* Inst = dyn_cast<Instruction>(User);
+      if (Inst->getOpcode() == Instruction::Load)
+        addFieldAccessNum(Inst, FieldLoc);
+      else if (Inst->getOpcode() == Instruction::Store){
+        if (U == Inst->getOperand(1))
+          addFieldAccessNum(Inst, FieldLoc);
+      }
+      else{
+        if (Inst->getOpcode() == Instruction::Call || Inst->getOpcode() == Instruction::Invoke){
+          addStats(StructFieldAccessManager::DebugStats::DS_GepPassedIntoFunc);
+        }
+        else if (Inst->getOpcode() == Instruction::BitCast){
+          addStats(StructFieldAccessManager::DebugStats::DS_GepPassedIntoBitcast);
+        }
+        else{
+          // TODO: Collect stats of this kind of access and add analysis later
+          addStats(StructFieldAccessManager::DebugStats::DS_GepUnknownUse, Inst->getOpcode());
+        }
+      }
+    }
+    else if (isa<Operator>(User)){
+      auto* Inst = dyn_cast<Operator>(U);
+      assert (Inst->getOpcode() != Instruction::Load || Inst->getOpcode() != Instruction::Store
+              || Inst->getOpcode() != Instruction::Call || Inst->getOpcode() != Instruction::Invoke);
+      if (Inst->getOpcode() == Instruction::BitCast){
+        addStats(StructFieldAccessManager::DebugStats::DS_GepPassedIntoBitcast);
+      }
+      else{
+        // TODO: Collect stats of this kind of access and add analysis later
+        addStats(StructFieldAccessManager::DebugStats::DS_GepUnknownUse, Inst->getOpcode());
+      }
+    }
+    else {
+      addStats(StructFieldAccessManager::DebugStats::DS_UserNotInstructionNorOperator);
+    }
+  }
+}
+
+void StructFieldAccessInfo::analyzeUsersOfStructValue(const Value* V)
+{
+  assert(V->getType()->isPointerTy() && V->getType()->getPointerElementType()->isStructTy());
+  for (auto *U : V->users()){
+    DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Analyzing user of " << *V << ": " << *U << "\n");
+    //assert(isa<Instruction>(U) || isa<Operator>(U));
+    if (isa<Instruction>(U)){
+      auto *Inst = dyn_cast<Instruction>(U);
+      if (Inst->getOpcode() != Instruction::GetElementPtr){
```

Comment:
Got it. I am just curious that is there any special reason not to use getOpcode()? Also, do I need to do the same thing for everywhere I check the opcode?